R/use_extendr.R

```diff
@@ -0,0 +1,170 @@
+    }
+    return(invisible(FALSE))
+  }
+
+  if (!isTRUE(quiet)) {
+    message(glue::glue("Creating `src` ..."))
+  }
+
+  dir.create(src_dir)
+  dir.create(file.path(src_dir, "rust"))
+  dir.create(file.path(src_dir, "rust", "src"))
+
+  entrypoint_content <- glue::glue(
+r"(
+// We need to forward routine registration from C to Rust
+// to avoid the linker removing the static library.
+
+void R_init_{pkg_name}_extendr(void *dll);
+
+void R_init_{pkg_name}(void *dll) {{
+    R_init_{pkg_name}_extendr(dll);
+}}
+)"
+  )
+  brio::write_lines(entrypoint_content, file.path(src_dir, "entrypoint.c"))
+
+  makevars_content <- glue::glue(
+"
+LIBDIR = ./rust/target/release
+STATLIB = $(LIBDIR)/lib{pkg_name}.a
+PKG_LIBS = -L$(LIBDIR) -l{pkg_name}
+
+all: C_clean
+
+$(SHLIB): $(STATLIB)
+
+$(STATLIB):
+\tcargo build --lib --release --manifest-path=./rust/Cargo.toml
+
+C_clean:
+\trm -Rf $(SHLIB) $(STATLIB) $(OBJECTS)
+
+clean:
+\trm -Rf $(SHLIB) $(STATLIB) $(OBJECTS) rust/target
+"
+  )
+  brio::write_lines(makevars_content, file.path(src_dir, "Makevars"))
+
+  makevars_win_content <- glue::glue(
+"
+TARGET = $(subst 64,x86_64,$(subst 32,i686,$(WIN)))-pc-windows-gnu
+LIBDIR = ./rust/target/$(TARGET)/release
+STATLIB = $(LIBDIR)/lib{pkg_name}.a
+PKG_LIBS = -L$(LIBDIR) -l{pkg_name} -lws2_32 -ladvapi32 -luserenv
+
+all: C_clean
+
+$(SHLIB): $(STATLIB)
+
+$(STATLIB):
+\tcargo build --target=$(TARGET) --lib --release --manifest-path=./rust/Cargo.toml
+
+C_clean:
+\trm -Rf $(SHLIB) $(STATLIB) $(OBJECTS)
+
+clean:
+\trm -Rf $(SHLIB) $(STATLIB) $(OBJECTS) rust/target
+"
+  )
+  brio::write_lines(makevars_win_content, file.path(src_dir, "Makevars.win"))
+
+  cargo_toml_content <- glue::glue(
+r"(
+  [package]
+  name = "{pkg_name}"
+  version = "0.1.0"
+  edition = "2018"
+
+  [lib]
+  crate-type = ["staticlib"]
+
+  [dependencies]
+  extendr-api = "*"
+)"
+  )
+  brio::write_lines(cargo_toml_content, file.path(src_dir, "rust", "Cargo.toml"))
+
+  lib_rs_content <- glue::glue(
+r"(
+use extendr_api::prelude::*;
+
+/// Return string `"Hello world!"` to R.
+/// @export
+#[extendr]
+fn hello_world() -> &'static str {{
+    "Hello world!"
+}}
+
+// Macro to generate exports.
+// This ensures exported functions are registered with R.
+// See corresponding C code in `entrypoint.c`.
+extendr_module! {{
+    mod {pkg_name};
+    fn hello_world;
+}}
+)"
+  )
+  brio::write_lines(lib_rs_content, file.path(src_dir, "rust", "src", "lib.rs"))
+
+
+  if (!isTRUE(quiet)) {
+    message(glue::glue("Creating `R/extendr-wrappers.R` ..."))
+  }
+
+
+  roxcmt <- "#'" # workaround for roxygen parsing bug in raw strings
+
+  wrappers_content <- glue::glue(
+r"(
+{roxcmt} @docType package
+{roxcmt} @usage NULL
+{roxcmt} @useDynLib {pkg_name}, .registration = TRUE
+NULL
+
+{roxcmt} Return string `"Hello world!"` to R.
+{roxcmt} @export
+hello_world <- function() .Call(wrap__hello_world)
+)"
+  )
+  brio::write_lines(wrappers_content, wrappers_file)
```

Comment:
Can we leave it to users to run `register_extendr()` instead of manually inserting these lines? I'm afraid we'll forget to fix here when we make some changes on wrapper generation...

Reply:
Users won't be able to compile unless they have the `@useDynLib` line. Therefore, I thought it would be best to add a wrapper file right away. It'll get overwritten the moment they rerun `register_extendr()`, so it only has to be approximately correct. (In fact, it just has to be not so wrong that the package won't build.)

Reply:
I see. Fair enough.